providers/fips/self_test_kats.c

```diff
@@ -193,6 +206,108 @@ static int self_test_kdf(const ST_KAT_KDF *t, OSSL_ST_EVENT *event,
+                                  int entropy, size_t min_len, size_t max_len,
+                                  int prediction_resistance)
+{
+    OSSL_PARAM *p = OSSL_PARAM_locate(drbg_params, DRBG_PARAM_ENTROPY);
+
+    if (p == NULL || p->data_type != OSSL_PARAM_OCTET_STRING)
+        return 0;
+    *pout = (unsigned char *)p->data;
+    return p->data_size;
+}
+
+static size_t drbg_kat_nonce_cb(RAND_DRBG *drbg, unsigned char **pout,
+                                int entropy, size_t min_len, size_t max_len)
+{
+    OSSL_PARAM *p = OSSL_PARAM_locate(drbg_params, DRBG_PARAM_NONCE);
+
+    if (p == NULL || p->data_type != OSSL_PARAM_OCTET_STRING)
+        return 0;
+    *pout = (unsigned char *)p->data;
+    return p->data_size;
+}
+
+static int self_test_drbg(const ST_KAT_DRBG *t, OSSL_ST_EVENT *event,
+                          OPENSSL_CTX *libctx)
+{
+    int ret = 0;
+    unsigned char out[256];
+    RAND_DRBG *drbg = NULL;
+    unsigned int flags = 0;
+    int prediction_resistance = 1; /* Causes a reseed */
+
+    SELF_TEST_EVENT_onbegin(event, OSSL_SELF_TEST_TYPE_DRBG, t->desc);
+
+    if (strcmp(t->desc, OSSL_SELF_TEST_DESC_DRBG_HMAC) == 0)
+        flags |= RAND_DRBG_FLAG_HMAC;
+
+    drbg = RAND_DRBG_new_ex(libctx, t->nid, flags, NULL);
+    if (drbg == NULL)
+        goto end;
+
+    if (!RAND_DRBG_set_callbacks(drbg, drbg_kat_entropy_cb, NULL,
+                                 drbg_kat_nonce_cb, NULL))
+        goto end;
+
+    drbg_params[0] =
+        OSSL_PARAM_construct_octet_string(DRBG_PARAM_ENTROPY,
+                                          (void *)t->entropyin, t->entropyinlen);
+    drbg_params[1] =
+        OSSL_PARAM_construct_octet_string(DRBG_PARAM_NONCE,
+                                          (void *)t->nonce, t->noncelen);
+
+    if (!RAND_DRBG_instantiate(drbg, t->persstr, t->persstrlen))
+        goto end;
+
+    drbg_params[0] =
+        OSSL_PARAM_construct_octet_string(DRBG_PARAM_ENTROPY,
+                                          (void *)t->entropyinpr1,
+                                          t->entropyinpr1len);
+
+    if (!RAND_DRBG_generate(drbg, out, t->expectedlen, prediction_resistance,
+                            t->entropyaddin1, t->entropyaddin1len))
+        goto end;
+
+    drbg_params[0] =
+        OSSL_PARAM_construct_octet_string(DRBG_PARAM_ENTROPY,
+                                         (void *)t->entropyinpr2,
+                                         t->entropyinpr2len);
+    /* This calls RAND_DRBG_reseed() internally when prediction_resistance = 1 */
+    if (!RAND_DRBG_generate(drbg, out,  t->expectedlen, prediction_resistance,
+                            t->entropyaddin2, t->entropyaddin2len))
+        goto end;
+
+    SELF_TEST_EVENT_oncorrupt_byte(event, out);
+
+    if (memcmp(out, t->expected, t->expectedlen) != 0)
+        goto end;
+
+    if (!RAND_DRBG_uninstantiate(drbg))
+        goto end;
+    /*
+     * TODO(3.0) : Check that the DRBG data has been zeroed after
```

Comment:
@mspncp : Can you investigate how to address this issue at some point?

Reply:
Huh, can you read my thoughts? I stumbled over your comment yesterday night and a good idea occurred to me, which kept turning around in my head all this day. I think I know now how to do it, but I didn't have the time to sit down and work it out on the computer.